haystack/document_stores/weaviate.py

```diff
@@ -337,30 +340,61 @@ def _get_current_properties(self, index: Optional[str] = None) -> List[str]:
 
         return cur_properties
 
-    def _build_filter_clause(self, filters: Dict[str, List[str]]) -> dict:
+    def _get_date_properties(self, index: Optional[str] = None) -> List[str]:
         """
-        Transform Haystack filter conditions to Weaviate where filter clauses.
+        Get all existing properties of type 'date' in the schema.
         """
-        weaviate_filters = []
-        weaviate_filter = {}
-        for key, values in filters.items():
-            for value in values:
-                weaviate_filter = {"path": [key], "operator": "Equal", "valueString": value}
-                weaviate_filters.append(weaviate_filter)
-        if len(weaviate_filters) > 1:
-            filter_dict = {"operator": "Or", "operands": weaviate_filters}
-            return filter_dict
-        else:
-            return weaviate_filter
+        index = self._sanitize_index_name(index) or self.index
+        cur_properties = []
+        for class_item in self.weaviate_client.schema.get()["classes"]:
+            if class_item["class"] == index:
+                cur_properties = [item["name"] for item in class_item["properties"] if item["dataType"][0] == "date"]
 
-    def _update_schema(self, new_prop: str, index: Optional[str] = None):
+        return cur_properties
+
+    def _update_schema(
+        self, new_prop: str, property_value: Union[List, str, int, float, bool], index: Optional[str] = None
+    ):
         """
         Updates the schema with a new property.
         """
         index = self._sanitize_index_name(index) or self.index
-        property_dict = {"dataType": ["string"], "description": f"dynamic property {new_prop}", "name": new_prop}
+        data_type = self._get_weaviate_type_of_value(property_value)
+
+        property_dict = {"dataType": [data_type], "description": f"dynamic property {new_prop}", "name": new_prop}
         self.weaviate_client.schema.property.create(index, property_dict)
 
+    @staticmethod
+    def _get_weaviate_type_of_value(value: Union[List, str, int, float, bool]) -> str:
+        """
+        Infers corresponding Weaviate data type for a value.
+        """
+        data_type = ""
+        list_of_values = False
+        if isinstance(value, list):
+            list_of_values = True
+            value = value[0]
+
+        if isinstance(value, str):
+            # If the value is parsable by datetime, it is a date
+            try:
+                datetime.fromisoformat(value)
```

Comment:
In filter_utils.py we have a similar method `_get_weaviate_datatype()`, which uses `value = convert_date_to_rfc3339(value)`. Could you please explain why we need two methods with this slightly different implementation?

Reply:
`_get_weaviate_datatype()` and `_get_weaviate_type_of_value()` are in fact quite similar. I would still keep them separate, as they have different purposes (and slightly different output).

`_get_weaviate_type_of_value()` is used when we dynamically create a new metadata field in Weaviate. We have to specify of what type the metadata field will be, which we infer from the value the user passes in. Here we can also have lists of str, int, ... as type. The resulting output is `"date"`, `"string"`, `"int"` "`number`" `"boolean"`, or, in case we have a list of values `"date[]"`, `"string[]"`, `"int[]"` "`number[]`" `"boolean[]"`.

`_get_weaviate_datatype()` is used for converting filters to weaviate format, as weaviate require to specify the type of the comparison value. The resulting output is one of `"valueDate"`, `"valueString"`, `"valueInt"`, `"valueNumber"`, `"valueBoolean"`. Lists are not allowed.

What do you think?

Reply:
Alright, thank you for the explanation. I am fine with keeping both methods. What I am still wondering is whether to change `datetime.fromisoformat(value)` to `convert_date_to_rfc3339(value)`. Imagine we change the implementation of `convert_date_to_rfc3339` in future so that it doesn't use `fromisoformat` anymore but something else. Changing the line of code here to `convert_date_to_rfc3339(value)` would make sure that this change doesn't need to be done here as well.